### PR TITLE
6248 add interfaces for rsync and rsync

### PR DESCRIPTION
**Pull request by Dwordcito**
|Related issue|
|---|
|#6248|

# Description
In order to do less trivial type conversions, a c ++ interface will be developed to consume the shared modules, since the new implementation of syscollector is in c ++.

## Interface's
image
image
image

## RTR
image

After Rework # 1
image

After Rework # 2
image

# DoD
- [x] Develop interfaces
- [x] Doxigen documentation
- [x] Run RTR
- [x] PR

### CODE REVIEW

File: src/shared_modules/dbsync/include/dbsync.h [hunk 1]
```diff
@@ -225,7 +225,7 @@ EXPORTED int dbsync_update_with_snapshot(const DBSYNC_HANDLE handle,
  * @return 0 if succeeded,
  *         specific error code (OS dependent) otherwise.
  */
-  EXPORTED int dbsync_update_with_snapshot_cb(const DBSYNC_HANDLE handle,
+EXPORTED int dbsync_update_with_snapshot_cb(const DBSYNC_HANDLE handle,
                                               const cJSON*        js_snapshot,
```
Comment (tinachoB) [inline src/shared_modules/dbsync/include/dbsync.h:229]: A little space indentation issue

File: src/shared_modules/dbsync/include/dbsync.hpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+     */
+    virtual void updateWithSnapshot(const nlohmann::json& jsInput,
+                                    ResultCallbackData&   callbackData);
+
+    /**
+     * @brief Turns off the services provided by the shared library.
+     */
+    static void teardown();
+
+    DBSYNC_HANDLE getHandle() { return m_dbsyncHandle; }
```
Comment (tinachoB) [inline src/shared_modules/dbsync/include/dbsync.hpp:144]: Missing API's doxygen comment

File: src/shared_modules/dbsync/include/dbsync.hpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+     * @param callbackData    Result callback(std::function) will be called for each result.
+     *
+     */
+    virtual void getDeletedRows(ResultCallbackData& callbackData);
+
+    TXN_HANDLE getHandle() { return m_txn; }
+
+private:
+    TXN_HANDLE m_txn;
+    bool m_shouldBeRemove;
```
Comment (tinachoB) [inline src/shared_modules/dbsync/include/dbsync.hpp:201]: m_shouldBeRemove -> m_shouldBeRemove**d**

File: src/shared_modules/dbsync/include/dbsync.hpp [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+    /**
+     * @brief Turns off the services provided by the shared library.
+     */
+    static void teardown();
+
+    DBSYNC_HANDLE getHandle() { return m_dbsyncHandle; } 
+private:
+    DBSYNC_HANDLE m_dbsyncHandle;
+    bool m_shouldBeRemove;
```
Comment (tinachoB) [inline src/shared_modules/dbsync/include/dbsync.hpp:147]: m_shouldBeRemove -> m_shouldBeRemoved

**Review by tinachoB — CHANGES_REQUESTED** [review]

**Review by tinachoB — APPROVED** [review]

File: src/shared_modules/dbsync/src/dbsync.cpp [hunk 5]
```diff
@@ -602,4 +605,161 @@ void dbsync_free_result(cJSON** js_data)
...
+#endif
+
+
+DBSync::DBSync(const HostType     hostType,
+               const DbEngineType dbType,
+               const std::string& path,
+               const std::string& sqlStatement)
+: m_shouldBeRemoved{ true }
+{
+    m_dbsyncHandle = DBSyncImplementation::instance().initialize(hostType, dbType, path, sqlStatement);
```
Comment (konangc87) [inline src/shared_modules/dbsync/src/dbsync.cpp:617]: move to initializer list.

File: src/shared_modules/dbsync/src/dbsync.cpp [hunk 6]
```diff
@@ -602,4 +605,161 @@ void dbsync_free_result(cJSON** js_data)
...
+            callbackData(result, jsonResult);
+        }
+    };
+    m_txn = PipelineFactory::instance().create(handle, tables, threadNumber, maxQueueSize, callbackWrapper);
+}
+
+DBSyncTxn::DBSyncTxn(const TXN_HANDLE handle)
+: m_shouldBeRemoved { false }
+{
+    m_txn = handle;
```
Comment (konangc87) [inline src/shared_modules/dbsync/src/dbsync.cpp:742]: ditto

File: src/shared_modules/rsync/src/rsync.cpp [hunk 7]
```diff
@@ -197,4 +198,64 @@ EXPORTED int rsync_close(const RSYNC_HANDLE handle)
...
+
+void RemoteSync::teardown()
+{
+    RSyncImplementation::instance().release();
+}
+
+RemoteSync::RemoteSync()
+: m_shouldBeRemoved{ true }
+{
+    m_handle = RSyncImplementation::instance().create();
```
Comment (konangc87) [inline src/shared_modules/rsync/src/rsync.cpp:211]: move to initializer list

**Review by konangc87 — CHANGES_REQUESTED** [review]

**Review by konangc87 — APPROVED** [review]

**Pull request merged**